sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java

```diff
@@ -546,41 +517,17 @@ public Registration(Class<T> proxyClass, List<PropertyDescriptor> beanInfo) {
    * @param iface The interface object to manually register.
    */
   public static synchronized void register(Class<? extends PipelineOptions> iface) {
-    checkNotNull(iface);
-    checkArgument(iface.isInterface(), "Only interface types are supported.");
-
-    if (REGISTERED_OPTIONS.contains(iface)) {
-      return;
-    }
-    validateWellFormed(iface, REGISTERED_OPTIONS);
-    REGISTERED_OPTIONS.add(iface);
+    CACHE.get().register(iface);
   }
 
   /**
    * Resets the set of interfaces registered with this factory to the default state.
    *
    * @see PipelineOptionsFactory#register(Class)
+   * @see Cache#Cache()
    */
-  @VisibleForTesting
-  static synchronized void resetRegistry() {
-    REGISTERED_OPTIONS.clear();
-    initializeRegistry();
-  }
-
-  /**
-   *  Load and register the list of all classes that extend PipelineOptions.
-   */
-  private static void initializeRegistry() {
-    register(PipelineOptions.class);
-    Set<PipelineOptionsRegistrar> pipelineOptionsRegistrars =
-        Sets.newTreeSet(ReflectHelpers.ObjectsClassComparator.INSTANCE);
-    pipelineOptionsRegistrars.addAll(
-        Lists.newArrayList(ServiceLoader.load(PipelineOptionsRegistrar.class, CLASS_LOADER)));
-    for (PipelineOptionsRegistrar registrar : pipelineOptionsRegistrars) {
-      for (Class<? extends PipelineOptions> klass : registrar.getPipelineOptions()) {
-        register(klass);
-      }
-    }
+  public static synchronized void resetCache() {
```

Comment:
Mark this `@Experimental` so if we think of a better way we can change it.

Also, why is that unloading the classloader that contains the PipelineOptionsFactory.class not a way to solve the memory leak since your already swapping class loaders.